### PR TITLE
[YUNIKORN-2393] Upgrade codecov/codecov-action to v4

### PR DESCRIPTION
**Pull request by chenyulin0719**
### What is this PR for?
Upgrade codecov/codecov-action from v3 to v4.
Support "secrete.CODECOV_TOKEN" for codecov analysis in forked repo.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
NA

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2393

### How should this be tested?
NA

### Screenshots (if appropriate)
NA

### Questions:
NA

### CODE REVIEW

**Comment by chenyulin0719** [comment]
Hi @chia7712,
In order to support build in our personal repo.
I added one more commit to set secrets.CODECOV_TOKEN.
(Everyone can follow the [doc](https://docs.codecov.com/docs/adding-the-codecov-token) to set codecov in their forked repo.)

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -37,6 +37,7 @@ jobs:
       - name: unit test
         run: make test
       - name: Code coverage
-        uses: codecov/codecov-action@v3
+        uses: codecov/codecov-action@v4
         with:
-         files: build/coverage.txt
+          files: build/coverage.txt
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (chia7712) [inline .github/workflows/main.yml:43]: @chenyulin0719 Please explain why we need this change? thanks!

Comment (chenyulin0719) [reply .github/workflows/main.yml:43]: Oh, I just put the comment above.
> Hi @chia7712,
In order to support build in our personal repo.
I added one more commit to set secrets.CODECOV_TOKEN.
(Everyone can follow the [doc](https://docs.codecov.com/docs/adding-the-codecov-token) to set codecov in their forked repo.)

Comment (chia7712) [reply .github/workflows/main.yml:43]: > In order to support build in our personal repo.
I added one more commit to set secrets.CODECOV_TOKEN.

got it. Could you add comments to the source code?

**Comment by chenyulin0719** [comment]
@chia7712, yes, I tested it in my personal repo.
-  https://github.com/chenyulin0719/yunikorn-core/pull/5

If we didn't set "CODECOV_TOKEN" in our personal repo,  this warning will pop out becasue no token was found.
image

File: .github/workflows/main.yml (hunk 1 above)
Comment (chenyulin0719) [reply .github/workflows/main.yml:43]: Sure.

**Comment by chenyulin0719** [comment]
Hi @chia7712, 
I just discovered that the report was uploaded, but there are no analysis results.
Let me figure out why.

**Comment by chenyulin0719** [comment]
Hi @chia7712,

I've validated the setting is correct and the Codecov Report was reported correctly in my forked repo too.
- https://github.com/chenyulin0719/yunikorn-core/pull/9#issuecomment-1932084825

**Comment by chia7712** [comment]
@chenyulin0719 Could you file PR to k8shim repo concurrently? I'd like to do double check for both repos :)

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request closed without merging**